fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/SimpleBufferBuilder.java

```diff
@@ -0,0 +1,456 @@
+            if (elementIndex != 0) throw new IllegalStateException("Not finished building vertex, Expected: " + format.types[elementIndex]); // You didn't finish building your vertex data.
+            if (vertices == 0) return 0; // No vertices buffered, lets not do anything.
+            if (vertices % mode.vertices != 0) throw new IllegalStateException("Does not contain vertices aligned to " + mode); // You did not put in enough vertices to cleanly slice the data into TRIANGLES/QUADS
+
+            // Reset position to 0, limit the buffer to our index.
+            buffer.position(0);
+            buffer.limit(index);
+
+            // Upload the raw vertex data in dynamic mode.
+            glBufferData(GL_ARRAY_BUFFER, buffer, GL_DYNAMIC_DRAW);
+
+            // The number of indices for triangles is equal to our vertex count, as that is
+            // what we operate in. However, for Quads, we have exactly vertices + vertices / 2
+            // vertices once we convert the quads to triangles.
+            indices = mode == Mode.TRIANGLES ? vertices : vertices + vertices / 2;
+
+            try (MemoryStack mStack = MemoryStack.stackPush()) {
+                // Allocate enough data to store all the indices we will need in bytes. We use ints for index.
+                ByteBuffer indexBuffer = mStack.malloc(indices * 4);
+                if (mode == Mode.TRIANGLES) {
+                    // Triangles are easy, just the number of vertices.
+                    for (int i = 0; i < vertices; i++) {
+                        indexBuffer.putInt(i);
+                    }
+                } else {
+                    int quads = vertices / 4;
+                    for (int i = 0; i < quads; i++) {
+                        // Quads are a bit different, we need to emit 2 triangles such that
+                        // when combined they make up a single quad.
+                        indexBuffer.putInt(i * 4 + 0).putInt(i * 4 + 1).putInt(i * 4 + 2);
+                        indexBuffer.putInt(i * 4 + 1).putInt(i * 4 + 3).putInt(i * 4 + 2);
+                    }
+                }
+                // Always flip position and limit!!
+                indexBuffer.flip();
+                // Upload the index buffer in dynamic mode.
+                glBufferData(GL_ELEMENT_ARRAY_BUFFER, indexBuffer, GL_DYNAMIC_DRAW);
+            }
+            return indices;
+        } finally {
+            // Reset builder state for next begin call.
+            building = false;
+            vertices = 0;
+            index = 0;
+        }
+    }
+
+    /**
+     * Upload and draw this buffer using one of a number of re-usable set of buffers.
+     * <p>
+     * This will immediately upload the buffer, resetting this builder for the next
+     * buffer operation, and draw the uploaded data.
+     * <p>
+     * You will need to bind shaders, textures, etc, before calling this function.
+     */
+    public void draw() {
```

Comment:
This style of drawing is extremely slow, if there is significant duplicate data from frame to frame, it should be pulled from on GPU. Re-using buffers like this also requires the driver to do more work and/or stall the GPU more. _Not that its really a performance issue during startup, but I'm still going to mention it_

Reply:
I'll bet it's faster than immediate mode draw calls tho 😂

Reply:
Would need to benchmark it, I wouldn't be too surprised either way.